[Docs] Batch07: adding examples to string functions (#16862)
apache/druid: docs/querying/sql-functions.md

## LOWER

Returns the expression in lowercase.

* **Syntax:** `LOWER(expr)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example converts the `OriginCityName` column from the `flight-carriers` datasource to lowercase.

```sql
SELECT 
  "OriginCityName" AS "original_string",

>   LOWER("OriginCityName") AS "make_lowercase"

Reviewer: ```suggestion
  LOWER("OriginCityName") AS "lowercase"
```

FROM "flight-carriers"
LIMIT 1
```

Returns the following:

> | `original_string` | `make_lowercase` |

Reviewer: ```suggestion
| `original_string` | `lowercase` |
```

| -- | -- |
`San Juan, PR` | `san juan, pr` |

</details>

[Learn more](sql-scalar.md#string-functions)

## LPAD

> Returns a string of size `length` from `expr`, left-padded with `chars`. The default value for `chars` is a space character. Truncates `expr` to `length` if `length` is shorter than the length of `expr`.

Reviewer: ```suggestion
Returns a string of size `length` from `expr`. When the length of `expr` is less than `chars`, left pads `expr` with `chars`, which defaults to the space character. Truncates `expr` to `length` if `length` is shorter than the length of `expr`.
```
Similarly for rpad

* **Syntax:** `LPAD(expr, length[, chars])`
* **Function type:** Scalar, string

<details><summary>Example</summary>

> The following example inserts `+` characters as padding to the left of the `OriginState` column from the `flight-carriers` datasource, making the return expression a length of 5 characters.

Reviewer: ```suggestion
The following example left pads the value of `OriginState` from `flight-carriers` to return a total of 5 characters.
```

```sql
SELECT 
  "OriginState" AS "origin_state",

>   LPAD("OriginState", 5, '+') AS "add_left_padding"

Reviewer 2: can we also add a TRUNCATE example?

Author: Extended the same example so that it would cover adding padding, truncating, and no change

## PARSE_LONG

Converts a string into a long(BIGINT) with the given radix, or into DECIMAL(base 10) if a radix is not provided.

* **Syntax:**`PARSE_LONG(string, [radix])`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example converts the string representation of the binary, radix 2, number `1100` into its long (BIGINT) equivalent.

```sql
SELECT 
  '1100' AS "binary_number_as_string", 
  PARSE_LONG('1110', 2) AS "parse_binary_string_into_BIGINT"
```

Returns the following:

> | `binary_number_as_string` | `parse_binary_string_into_BIGINT` |

Reviewer: you haven't been in the habit of using any caps in the output columns, so this feels inconsistent. also I'd encourage you to look at a way to be a hint more concise in the names if possible. would `binary_as_string` and `bigint_value` work? What about for the other functions with long output columns?

## POSITION

Returns the one-based index position of a substring within an expression, optionally starting from a given one-based index. If `substring` is not found, returns 0.

* **Syntax**: `POSITION(substring IN expr [FROM startingIndex])`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example returns the one-based index of the substring `PR` in the `OriginCityName` column from the `flight-carriers` datasource starting from index 5.

```sql
SELECT 
  "OriginCityName" AS "original_expr",

>   POSITION('PR' IN "OriginCityName" FROM 5) AS "index_found"

Reviewer: `index_found` makes me think of a boolean. this actually returns the index. `index` is probably enough. if it is a 1-based index 0 is not in range, so understandably, not found.

## REGEXP_EXTRACT

Apply regular expression `pattern` to `expr` and extract the `N`-th capture group. If `N` is unspecified or zero, returns the first substring that matches the pattern. Returns `null` if there is no matching pattern.

* **Syntax:** `REGEXP_EXTRACT(expr, pattern[, N])`
* **Function type:** Scalar, string 

<details><summary>Example</summary>

The following example uses regular expressions to find city names inside the `OriginCityName` column from the `flight-carriers` datasource by matching what comes before the comma.

```sql
SELECT 
  "OriginCityName" AS "original_expr",
  REGEXP_EXTRACT("OriginCityName", '([^,]+)', 0) AS "pattern_extracted"
FROM "flight-carriers"
LIMIT 1
```

Returns the following:

> | `original_expr` | `pattern_extracted` |

Reviewer: just `pattern` ?

Author: When I think `pattern` I think of the regular expression, `([^,]+)`. I do think `pattern_extracted` is more appropriate. But maybe we can consider `extracted` or `pattern_match`

Author: `pattern_match`

| -- | -- |
| `San Juan, PR` | `San Juan`|

</details>

[Learn more](sql-scalar.md#string-functions)

## REGEXP_LIKE

Returns `true` if the regular expression `pattern` finds a match in `expr`. Returns `false` otherwise.

* **Syntax:** `REGEXP_LIKE(expr, pattern)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

> The following example returns `true` when the `OriginCityName` column from `flight-carriers` has a city name with two words.

Reviewer: Would "city name with words separated by a space" be more accurate?

Author: Yeah I think yours is more accurate, but what about "city name with a space in the name"?

```sql
SELECT 
  "OriginCityName" AS "origin_city",
  REGEXP_LIKE("OriginCityName", '[A-Za-z]+\s[A-Za-z]+') AS "pattern_found"
FROM "flight-carriers"
LIMIT 2
```

Returns the following:

| `origin_city` | `pattern_found` |
| -- | -- |
| `San Juan, PR` | `true` |
| `Boston, MA` | `false` |

</details>

[Learn more](sql-scalar.md#string-functions)

## REGEXP_REPLACE

Replaces all occurrences of a regular expression in a string expression with a replacement string. The replacement string may refer to capture groups using `$1`, `$2`, etc.

* **Syntax:** `REGEXP_REPLACE(expr, pattern, replacement)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example matches instances of the word `Fort` and replaces it with it's abbreviation `Ft.`

```sql
SELECT 
  'Fort Lauderdale, FL' AS "original_string",

>   REGEXP_REPLACE('Fort Lauderdale, FL', 'Fort', 'Ft.') AS "modified_string"

Reviewer 2: can we also add an example of how capture group replacement(like $1) is used?

Author: changes the example entirely to make use of capture groups

```

Returns the following:

> | `original_expr` | `replace_fort_with_abbreviation` |

Reviewer: consider different column aliases. Have you been using abbreviations like "expr" regularly or have you been using the words spelled out?

Author: changed it to  `original_stringr` and `replace_pattern`

## RPAD

Returns a string of size `length` from `expr`, right-padded with `chars`. The default value for `chars` is a space character. Truncates `expr` to `length` if `length` is shorter than the length of `expr`.

* **Syntax:** `RPAD(expr, length[, chars])`
* **Function type:** Scalar, string

<details><summary>Example</summary>

> The following example inserts `+` characters as padding to the right of the `OriginState` column from the `flight-carriers` datasource, making the return expression a length of `5` characters.

Reviewer: why does the length of the return string matter?

Author: I included it because the definition includes: "Returns a string of size `length`", so I though it important to point it out in the definition. I do not think it is absolutely necessary, but I still think its useful to distinguish that its different than just insert x characters to the front/back of the string

Author: Ignoring

```sql
SELECT 
  "OriginState" AS "origin_state",

>   RPAD("OriginState", 5, '+') AS "add_right_padding"

Reviewer 2: can we also add a TRUNCATE example?

Author: Extended the same example so that it would cover adding padding, truncating, and no change

## UPPER

Returns the expression in uppercase.

* **Syntax:** `UPPER(expr)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example converts the `OriginCityName` column from the `flight-carriers` datasource to uppercase.

```sql
SELECT 
  "OriginCityName" AS "original_string",
  UPPER("OriginCityName") AS "make_uppercase"
FROM "flight-carriers"
LIMIT 1
```

Returns the following:

> | `original_string` | `make_uppercase` |

Reviewer: `origin_city`, `uppercase`

## REGEXP_EXTRACT

Apply regular expression `pattern` to `expr` and extract the Nth capture group. If `N` is unspecified or zero, returns the first substring that matches the pattern. Returns `null` if there is no matching pattern.

* **Syntax:** `REGEXP_EXTRACT(expr, pattern[, N])`
* **Function type:** Scalar, string 

<details><summary>Example</summary>

The following example uses regular expressions to find city names inside the `OriginCityName` column from the `flight-carriers` datasource by matching what comes before the comma.

```sql
SELECT 
  "OriginCityName" AS "origin_city",
  REGEXP_EXTRACT("OriginCityName", '([^,]+)', 0) AS "pattern_match"
FROM "flight-carriers"
LIMIT 1
```

Returns the following:

| `origin_city` | `pattern_match` |
| -- | -- |
| `San Juan, PR` | `San Juan`|

</details>

[Learn more](sql-scalar.md#string-functions)

## REGEXP_LIKE

Returns `true` if the regular expression `pattern` finds a match in `expr`. Returns `false` otherwise.

* **Syntax:** `REGEXP_LIKE(expr, pattern)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

> The following example returns `true` when the `OriginCityName` column from `flight-carriers` has a city name with a space in the name.

Reviewer: ```suggestion
The following example returns `true` when the `OriginCityName` column from `flight-carriers` has a city name containing a space.
```

```sql
SELECT 
  "OriginCityName" AS "origin_city",
  REGEXP_LIKE("OriginCityName", '[A-Za-z]+\s[A-Za-z]+') AS "pattern_found"
FROM "flight-carriers"
LIMIT 2
```

Returns the following:

| `origin_city` | `pattern_found` |
| -- | -- |
| `San Juan, PR` | `true` |
| `Boston, MA` | `false` |

</details>

[Learn more](sql-scalar.md#string-functions)

## REGEXP_REPLACE

> Replaces all occurrences of a regular expression in a string expression with a replacement string. The replacement string may refer to capture groups using `$1`, `$2`, etc.

Reviewer: ```suggestion
Replaces all occurrences of a regular expression in a string expression with a replacement string. Refer 
 to capture groups in the replacement string using `$group` syntax. For example: `$1` or `$2`.
```

* **Syntax:** `REGEXP_REPLACE(expr, pattern, replacement)`
* **Function type:** Scalar, string

<details><summary>Example</summary>

The following example matches three consecutive words, where each word is its own capture group, and replaces the matched words with the word in the second capture group punctuated with exclamation marks.

```sql
SELECT 
  'foo bar baz' AS "original_string",

>   REGEXP_REPLACE('foo bar baz', '([A-Za-z]+) ([A-Za-z]+) ([A-Za-z]+)' , '$2!!') AS "modified_string"

Reviewer: ```suggestion
  REGEXP_REPLACE('foo bar baz', '([A-Za-z]+) ([A-Za-z]+) ([A-Za-z]+)' , '$2!') AS "modified_string"
```
the way it was, the result was `bar!!`
